contracts/v0.7/bridge/optimism/OVM_L1ERC20Gateway.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: MIT
+// @unsupported: ovm
+pragma solidity >0.6.0 <0.8.0;
+pragma experimental ABIEncoderV2;
+
+/* Interface Imports */
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/* Library Imports */
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+
+/* Contract Imports */
+import { Abs_L1TokenGateway } from "@eth-optimism/contracts/dist/contracts/OVM/bridge/tokens/Abs_L1TokenGateway.sol";
+import { Initializable } from "@openzeppelin/contracts/proxy/Initializable.sol";
+import { OpUnsafe } from "../utils/OpUnsafe.sol";
+
+/**
+ * @title OVM_L1ERC20Gateway
+ * @dev The L1 ERC20 Gateway is a contract which stores deposited L1 funds that are in use on L2.
+ * It synchronizes a corresponding L2 ERC20 Gateway, informing it of deposits, and listening to it
+ * for newly finalized withdrawals.
+ *
+ * NOTE: This contract extends Abs_L1TokenGateway, which is where we
+ * takes care of most of the initialization and the cross-chain logic.
+ * If you are looking to implement your own deposit/withdrawal contracts, you
+ * may also want to extend the abstract contract in a similar manner.
+ *
+ * Compiler used: solc
+ * Runtime target: EVM
+ */
+contract OVM_L1ERC20Gateway is OpUnsafe, Initializable, Abs_L1TokenGateway {
+  // L1 token we are bridging to L2
+  IERC20 public s_l1ERC20;
+
+  // This contract lives behind a proxy, so the constructor parameters will go unused.
+  constructor()
+    Abs_L1TokenGateway(
+      address(0), // _l2DepositedToken
+      address(0) // _l1messenger
+    )
+    public
+  {}
+
+  /**
+   * @param l2ERC20Gateway L2 Gateway address on the chain being deposited into
+   * @param l1Messenger Cross-domain messenger used by this contract.
+   * @param l1ERC20 L1 ERC20 address this contract stores deposits for
+   */
+  function init(
+    address l2ERC20Gateway,
+    address l1Messenger,
+    IERC20 l1ERC20
+  )
+    initializer()
+    public
+  {
+    // TODO: require != address(0)?
```

Comment:
seems like a good idea. since `init` can only be called once because of the `initializer`, would be a good idea to make sure that none of the addresses are set to `address(0)` by accident